EventFilter/CTPPSRawToDigi/src/CTPPSPixelDataFormatter.cc

```diff
@@ -124,63 +124,75 @@ void CTPPSPixelDataFormatter::interpretRawData(
     mit = m_Mapping.find(fPos);
-
-    if (mit == m_Mapping.end()) {
-      if (nlink >= maxLinkIndex) {
-        m_ErrorCheck.conversionError(fedId, iD, InvalidLinkId, ww, errors);
-        edm::LogError("CTPPSPixelDataFormatter") << " Invalid linkId ";
-      } else if ((nroc - 1) >= maxRocIndex) {
-        // if RocId wrong try to recover at least plane id
-        CTPPSPixelFramePosition fPosDummyROC(fedId, FMC, nlink, 0);
-        mit = m_Mapping.find(fPosDummyROC);
-        if (mit != m_Mapping.end())
-          iD = (*mit).second.iD;
-        m_ErrorCheck.conversionError(fedId, iD, InvalidROCId, ww, errors);
-        edm::LogError("CTPPSPixelDataFormatter")
-            << " Invalid ROC Id " << convroc << " in nlink " << nlink << " of FED " << fedId << " in DetId " << iD;
-      } else {
-        m_ErrorCheck.conversionError(fedId, iD, Unknown, ww, errors);
-        edm::LogError("CTPPSPixelDataFormatter") << " Error unknown ";
-      }
-      continue;  //skip word
+    if (mit != m_Mapping.end()) {
+      CTPPSPixelROCInfo rocInfo = (*mit).second;
+      iD = rocInfo.iD;
+      rocp.setParameters(iD, rocInfo.roc, convroc);
     }
 
-    CTPPSPixelROCInfo rocInfo = (*mit).second;
-    iD = rocInfo.iD;
-    CTPPSPixelROC rocp(iD, rocInfo.roc, convroc);
-
     if ((nlink != link) | (nroc != roc)) {  // new roc
       link = nlink;
       roc = nroc;
 
-      skipROC = LIKELY((roc - 1) < maxRocIndex) ? false : !m_ErrorCheck.checkROC(errorsInEvent, fedId, iD, ww, errors);
+      if ((roc - 1) < maxRocIndex) {
+        skipROC = false;
+      } else {
+        // using dummy detId - recovering of FED channel foreseen in DQM
+        iD = RPixErrorChecker::dummyDetId;
+        skipROC = !m_ErrorCheck.checkROC(errorsInEvent, fedId, iD, ww, errors);
+      }
       if (skipROC)
         continue;
 
+      if (mit == m_Mapping.end()) {
+        if (nlink >= maxLinkIndex) {
+          m_ErrorCheck.conversionError(fedId, iD, InvalidLinkId, ww, errors);
+          edm::LogError("CTPPSPixelDataFormatter") << " Invalid linkId ";
+        } else if ((nroc - 1) >= maxRocIndex) {
+          m_ErrorCheck.conversionError(fedId, iD, InvalidROCId, ww, errors);
+          edm::LogError("CTPPSPixelDataFormatter")
+              << " Invalid ROC Id " << convroc << " in nlink " << nlink << " of FED " << fedId << " in DetId " << iD;
+        } else {
+          m_ErrorCheck.conversionError(fedId, iD, Unknown, ww, errors);
+          edm::LogError("CTPPSPixelDataFormatter") << " Error unknown ";
+        }
+        skipROC = true;  // skipping roc due to mapping errors
+        continue;
+      }
+      if (rocp.rawId() == 0) {
+        skipROC = true;
+        continue;
+      }
+
       auto rawId = rocp.rawId();
 
       detDigis = &digis.find_or_insert(rawId);
       if ((*detDigis).empty())
         (*detDigis).data.reserve(32);  // avoid the first relocations
     }
 
+    if (skipROC || rocp.rawId() == 0)
```

Comment:
@fabferro , just for checking: is it correct that `skipROC` is defined outside the loop, i.e. it remains set at the value taken during the previous iterations in the loop over `word`'s?

Reply:
Thanks Andrea for the question. The use of this flag is very delicate. The way it's assumed to work is that if it happens to be set true (i.e. a roc chip behaving badly) all the words from that roc are skipped. The flag can change inside the loop when the roc number changes (see lines 156-162) and the new is roc checked.